sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/Query13.java

```diff
@@ -0,0 +1,107 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.nexmark.queries;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Random;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.nexmark.NexmarkConfiguration;
+import org.apache.beam.sdk.nexmark.model.Event;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Reshuffle.AssignShardFn;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * Query "13" PORTABILITY_BATCH (not in original suite).
+ *
+ * <p>This benchmark is created to stress the boundary of runner and SDK in portability world. The
+ * basic shape of this benchmark is source + GBK + ParDo, in which the GBK read + ParDo will require
+ * that runner reads from shuffle and connects with SDK to do CPU intensive computation.
+ */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class Query13 extends NexmarkQueryTransform<Event> {
+  private final NexmarkConfiguration configuration;
+
+  public Query13(NexmarkConfiguration configuration) {
+    super("Query13");
+    this.configuration = configuration;
+  }
+
+  @Override
+  public PCollection<Event> expand(PCollection<Event> events) {
+    final Coder<Event> coder = events.getCoder();
+    return events
+        .apply("Pair with random key", ParDo.of(new AssignShardFn<>(configuration.numKeyBuckets)))
+        .apply(GroupByKey.create())
+        .apply(
+            "ExpandIterable",
+            ParDo.of(
+                new DoFn<KV<Integer, Iterable<Event>>, Event>() {
+                  @ProcessElement
+                  public void processElement(
+                      @Element KV<Integer, Iterable<Event>> element, OutputReceiver<Event> r) {
+                    for (Event value : element.getValue()) {
+                      r.output(value);
+                    }
+                  }
+                }))
+        // Force round trip through coder.
+        .apply(
+            name + ".Serialize",
+            ParDo.of(
+                new DoFn<Event, Event>() {
+                  private final Counter bytesMetric = Metrics.counter(name, "serde-bytes");
+                  private final Random random = new Random();
+                  private double pardoCPUFactor =
+                      (configuration.pardoCPUFactor >= 0.0 && configuration.pardoCPUFactor <= 1.0)
```

Comment:
I think the condition check is probably redundant

Reply:
I prefer to treat a value outside of 0.0~1.0 to default value because value outside this range seems less meaningful. Though I guess without this check the code will still work.